compiler/rustc_mir_build/src/build/expr/as_place.rs

```diff
@@ -70,51 +72,148 @@ struct PlaceBuilder<'tcx> {
     projection: Vec<PlaceElem<'tcx>>,
 }
 
-fn capture_matching_projections<'a, 'tcx>(
-    tcx: TyCtxt<'tcx>,
+/// Given a list of MIR projections, convert them to list of hir ProjectionKind.
+fn convert_to_hir_projections_for_capture<'tcx>(
+    mir_projections: &Vec<PlaceElem<'tcx>>) -> Vec<HirProjectionKind> {
+
+    let mut hir_projections  = Vec::new();
+
+    for mir_projection in mir_projections {
+        let hir_projection = match mir_projection {
+            ProjectionElem::Deref => HirProjectionKind::Deref,
+            ProjectionElem::Field(field, _) => {
+                // We will never encouter this for multivariant enums,
+                // read the comment for `Downcast`.
+                HirProjectionKind::Field(field.index() as u32, VariantIdx::new(0))
+            },
+            ProjectionElem::Downcast(..) => {
+                // This projections exist only for enums that have
+                // multiple variants. Since such enums that are captured
+                // completely, we can stop here.
+                break
+            },
+            ProjectionElem::Index(..)
+            | ProjectionElem::ConstantIndex { .. }
+            | ProjectionElem::Subslice { .. } => {
+                // We don't capture array-access projections.
+                // We can stop here as arrays are captured completely.
+                break
+            },
+        };
+
+        hir_projections.push(hir_projection);
+    }
+
+    hir_projections
+}
+
+/// Return true if the `proj_possible_ancestor` represents an ancestor path
+/// to `proj_capture`, assuming they both start off of the same root variable.
+///
+/// Eg: `foo.x` which is represented using `projections=[Field(x)]` is an ancestor of
+///     `foo.x.y` which is represented using `prjections=[Field(x), Field(y)]`.
+///     Note both `foo.x` and `foo.x.y` start off of the same root variable `foo`.
+fn is_ancestor_of_capture(
+    proj_possible_ancestor: &Vec<HirProjectionKind>,
+    proj_capture: &Vec<HirProjectionKind>) -> bool {
+
+    for (proj_ans, proj_capture) in proj_possible_ancestor.iter().zip(proj_capture.iter()) {
+        if proj_ans != proj_capture {
+            return false;
+        }
+    }
+
+    // In cases like
+    // ```rust,ignore(wont-compile-because-pseudocode)
+    // let c = ||  {
+    //     x.0.1 += 1;
+    //     let _ = x.0;
+    // }
+    // ```
+    //
+    // Here with `capture_disjoint_fields` enabled, x.0 won't be
+    // captured but x.0.1 would be.
+    //
+    // Therefore we can't just return true here.
+    proj_possible_ancestor.len() <= proj_capture.len()
+}
+
+/// Computes the index of a capture within the desugared closure provided the closure's
+/// `closure_min_captures` and the capture's index of the capture in the
+/// `ty::MinCaptureList` of the root variable `var_hir_id`.
+fn compute_capture_idx<'tcx>(
```

Comment:
This can probably be optimized by moving it into the min_capture map itself, so we move the table from looking like
```
root variable -> list of captures
```
to 

```
root variarble -> { start_idx, list of captures }
```

where the `start_idx` is essentially the index in `min_capture` map for the first capture in the list of captures for the root variable.